lib/storage/src/content_manager/shard_distribution.rs

```diff
@@ -0,0 +1,98 @@
+use collection::shard::{Shard, ShardId};
+use collection::PeerId;
+use serde::{Deserialize, Serialize};
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+struct PeerShardCount {
+    shard_count: usize, // self.shard_count and other.shard_count are compared first to determine eq & ord
+    peer_id: PeerId,
+}
+
+impl PeerShardCount {
+    fn new(shard_count: usize, peer_id: PeerId) -> Self {
+        Self {
+            shard_count,
+            peer_id,
+        }
+    }
+
+    fn inc_shard_count(&mut self) {
+        self.shard_count += 1;
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Hash)]
+pub struct ShardDistributionProposal {
+    pub distribution: Vec<(ShardId, PeerId)>,
+}
+
+impl ShardDistributionProposal {
+    /// Builds a proposal for the distribution of shards.
+    /// It will propose to allocate shards so that all peers have the same number of shards at the end.
+    pub fn build(
+        config_shard_number: u32,
+        this_peer_id: PeerId,
+        known_peers: &[PeerId],
+        known_shards: Vec<&Shard>,
+    ) -> Self {
```

Comment:
Same comment about `build` here. But generally it is just a convention, so if there are good reasons for this naming - I don't mind